asset/asset.go

```diff
@@ -943,6 +944,447 @@ type GroupKeyReveal interface {
+		version:     1,
+		internalKey: ToSerialized(&internalKey),
+		tapscript:   gkrTapscript,
+	})
+}
+
+// ScriptSpendControlBlock returns the control block for the script spending
+// path in the custom tapscript subtree.
+//
+// nolint: lll
+func (g *GroupKeyRevealV1) ScriptSpendControlBlock(
+	genesisAssetID ID) lfn.Result[txscript.ControlBlock] {
+
+	internalKey, err := btcec.ParsePubKey(g.internalKey[:])
+	if err != nil {
+		return lfn.Errf[txscript.ControlBlock]("failed to parse "+
+			"internal key: %w", err)
+	}
+
+	outputKey := txscript.ComputeTaprootOutputKey(
+		internalKey, g.tapscript.root[:],
+	)
+	outputKeyIsOdd := outputKey.SerializeCompressed()[0] == 0x03
+
+	// If the custom subtree inclusion proof is nil, it may not have been
+	// set during decoding. Compute it, set it on the group key reveal, and
+	// validate both the computed tapscript root against the expected
+	// root.
+	if len(g.tapscript.customSubtreeInclusionProof) == 0 {
+		gkrTapscript, err := NewGroupKeyTapscriptRoot(
+			genesisAssetID, g.tapscript.customSubtreeRoot,
+		).Unpack()
+		if err != nil {
+			return lfn.Errf[txscript.ControlBlock]("failed to "+
+				"generate tapscript artifacts: %w", err)
+		}
+
+		// Ensure that the computed tapscript root matches the expected
+		// root.
+		if !gkrTapscript.root.IsEqual(&g.tapscript.root) {
+			return lfn.Errf[txscript.ControlBlock]("tapscript "+
+				"root mismatch (expected=%s, computed=%s)",
+				g.tapscript.root, gkrTapscript.root)
+		}
+
+		// Set the custom subtree inclusion proof on the group key
+		// reveal.
+		g.tapscript.customSubtreeInclusionProof =
+			gkrTapscript.customSubtreeInclusionProof
+	}
+
+	return lfn.Ok(txscript.ControlBlock{
+		InternalKey:     internalKey,
+		OutputKeyYIsOdd: outputKeyIsOdd,
+		LeafVersion:     txscript.BaseLeafVersion,
+		InclusionProof:  g.tapscript.customSubtreeInclusionProof,
+	})
+}
+
+// Encode encodes the group key reveal into a writer.
+//
+// This encoding routine must ensure the resulting serialized bytes are
+// sufficiently long to prevent the decoding routine from mistakenly using the
+// wrong group key reveal version. Specifically, the raw key, tapscript root,
+// and version fields must be properly populated.
+func (g *GroupKeyRevealV1) Encode(w io.Writer) error {
+	records := []tlv.Record{
+		NewGKRVersionRecord(&g.version),
+		NewGKRInternalKeyRecord(&g.internalKey),
+		NewGKRTapsciptRootRecord(&g.tapscript.root),
+	}
+
+	// Add encode record for the custom tapscript root, if present.
+	g.tapscript.customSubtreeRoot.WhenSome(func(hash chainhash.Hash) {
+		records = append(records, NewGKRCustomSubtreeRootRecord(&hash))
+	})
+
+	stream, err := tlv.NewStream(records...)
+	if err != nil {
+		return err
+	}
+	return stream.Encode(w)
+}
+
+// Decode decodes the group key reveal from a reader.
+func (g *GroupKeyRevealV1) Decode(r io.Reader, buf *[8]byte, l uint64) error {
+	var customSubtreeRoot chainhash.Hash
+
+	tlvStream, err := tlv.NewStream(
```

Comment:
Ah ok, I see now the decomposition of the reveal struct, and the auxiliary tapscript proof information.